solidity/contracts/libs/middleware/InterchainQueryMessage.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import {CallLib} from "../Call.sol";
+
+/**
+ * Format of message:
+ * [   0: 32] Sender address
+ * [  32: 64] Message type (left padded with zeroes)
+ * [  64:???] Encoded call array
+ */
+library InterchainQueryMessage {
+    uint256 private constant SENDER_OFFSET = 0;
+    uint256 private constant TYPE_OFFSET = 32;
+    uint256 private constant CALLS_OFFSET = 64;
+
+    enum MessageType {
+        QUERY,
+        RESPONSE
+    }
+
+    /**
+     * @notice Parses and returns the query sender from the provided message
+     * @param _message The interchain query message
+     * @return The query sender as bytes32
+     */
+    function sender(bytes calldata _message) internal pure returns (bytes32) {
+        return bytes32(_message[SENDER_OFFSET:TYPE_OFFSET]);
+    }
+
+    /**
+     * @notice Parses and returns the message type from the provided message
+     * @param _message The interchain query message
+     * @return The message type (query or response)
+     */
+    function messageType(bytes calldata _message)
+        internal
+        pure
+        returns (MessageType)
+    {
+        // left padded with zeroes
+        return MessageType(uint8(bytes1(_message[CALLS_OFFSET - 1])));
+    }
+
+    /**
+     * @notice Returns formatted InterchainQueryMessage, type == QUERY
+     * @param _sender The query sender as bytes32
+     * @param _calls The sequence of queries to make, with the corresponding
+     * response callbacks
+     * @return Formatted message body
+     */
+    function format(
+        bytes32 _sender,
+        CallLib.StaticCallWithCallback[] calldata _calls
+    ) internal pure returns (bytes memory) {
+        return abi.encode(_sender, MessageType.QUERY, _calls);
```

Comment:
@asaj why isn't the encoding similar to and as efficient as `InterchainAccountMessage` being used here

Fore example `format` :
```bash
function format(
    bytes32 _sendar,
    CallLib.StaticCallWithCallback[] calldata _calls
){
    return abi.encodePacked(_sender, types, abi.encode(calls));
}
```

Decoding this is as simple as

```bash
        _sender = bytes32(data[0:32]);
        types = MessageTypes(uint8(data[32]));
        calls = abi.decode(data[33:], (StaticCallWithCallback[]));
```

Reply:
I think @yorhodes correctly called out that this is not as gas efficient as just `abi.encode()`